ASP-based solver: allow to reuse installed externals

alalazo opened the pull request:
fixes #31484
fixes #31832

Before this change if anything was matching an external condition, it was considered "external" and thus something to be "built". This was happening in particular to external packages that were re-read from the DB, which then couldn't be reused, causing the problems shown in #31484.

This PR fixes the issue by excluding specs with a "hash" from being considered "external"

alalazo committed:
ASP-based solver: allow to reuse installed externals

fixes #31484

Before this change if anything was matching an external
condition, it was considered "external" and thus something
to be "built".

This was happening in particular to external packages
that were re-read from the DB, which then couldn't be
reused, causing the problems shown in #31484.

This PR fixes the issue by excluding specs with a
"hash" from being considered "external"

becker33 commented on lib/spack/spack/solver/concretize.lp:
```diff
@@ -459,7 +459,9 @@ version_weight(Package, Weight) :- external_version(Package, Version, Weight).
 
 % if a package is not buildable (external_only), only externals are allowed
-external(Package) :- external_only(Package), node(Package).
+external(Package) :- external_only(Package),
```
I think this will allow you to reuse a non-external even when `buildable: false` for that package. Can you add a test for this?

alalazo replied:
I don't think this is wrong. It will allow what you say only if the non-external is already installed, so not buildable, but built. Disallowing that case causes the bug in this way:
1. Install an external that is not buildable
2. Can't reuse the same external at the next concretization

adamjstewart reviewed (dismissed):
Works like a charm, thanks for the quick fix!

Also tested that the tests do indeed fail without this change.

alalazo commented on lib/spack/spack/test/concretize.py:
```diff
@@ -1836,3 +1836,33 @@ def test_installed_externals_are_reused(
             external3 = Spec('changing@1.0').concretized()
         assert external3.dag_hash() == external1.dag_hash()
+
+    @pytest.mark.regression('31484')
+    def test_installed_builds_not_reused_buildable_false(
+            self, mutable_database, repo_with_changing_recipe
+    ):
```
@becker33 I have been getting back at this specific test, and I think this is behavior we don't want to add as it is coded here. The only things we  currently say for `buildable: false` in our documentation are:

> The addition of the buildable flag tells Spack that it should never build its own version of OpenMPI, and it will instead always rely on a pre-built OpenMPI.
>
> The buildable does not need to be paired with external packages. It could also be used alone to forbid packages that may be buggy or otherwise undesirable.
> [ ... ]

and they clearly stem from a world where `reuse` was not a considered at all. It seems thus a stretch to me to require that `buildable: false` prevents the reuse of software that is either installed locally or available from a buildcache.

That said, I concur, based on our discussion on Slack, that there's a gap in expressiveness that we need to fill. For instance, as you pointed out:
```yaml
packages:
  mpi:
    buildable: false
```
likely means that the user wants to install only from one of the externals that might be listed in the same file. But what if it is:
```yaml
packages:
  trilinos:
    buildable: false
```
Couldn't that be a way to say "Reuse a binary version of trilinos and never build that from sources"? I wouldn't find that request "crazy", if a binary cache is available. 

I think the point is: `buildable: false` was added before `reuse` and we didn't figure out how the two should interact together. That is possibly what needs to get fixed (allow the user to express in which of the two cases they are), not excluding specs from reuse based on `buildable: false`.

**TL;DR** I agree that here we miss ways to express a few requests, but would it be possible to get the bugfix in as is, and work on a more extensive feature in a following PR? lf not I'm tempted to move this out of the v0.18.1 project, since it likely means it won't be solved soon.

alalazo replied:
> I still don't see how someone using this is supposed to use system mpi, and I don't think that's a use case we can afford to break.

@becker33 Fair, I already moved this to the next bugfix release. Do you agree though that the test, as is, is not something we want to force implicitly?

sethrj commented:
@alalazo I think this will also fix #31832 . Interestingly my `libuuid` external is a build-only dependency of the spec that's failing... so it doesn't appear in `spack find -lvd`. However I guess it still gets used by the concretizer?

adamjstewart commented:
@becker33 @alalazo any updates on this? Spack is basically unusable on macOS if you update develop frequently. Not sure if there's a priority level higher than "impact-high" lol. I've been using this PR locally on a couple systems since this PR was first opened and I would like to no longer have to use that local modification.

alalazo commented:
@adamjstewart There is currently an issue, that @becker33 pointed out, that is hard to solve in this PR alone. My thoughts on this are:
1. The second test added is imho not what we want from the perspective of the semantics
2. I am trying to push forward the review of #27987 to have hard requirements in `packages.yaml`
3. When #27987 is merged, we can remove the second test here (maybe add another one to ensure we don't have regression on the issue with `mpi:buildable:false`) and merge this

Rationale is that, if we have requirements, we can specify there what needs to be used.

sethrj commented:
@alalazo @adamjstewart @becker33 is there a quick fix for getting this in for macos platforms?

alalazo commented:
@sethrj https://github.com/spack/spack/pull/27987 will be merged today or tomorrow. With that in I plan to:
1. Rebase this PR
2. Substitute the last test with a similar one that uses requirements
3. :crossed_fingers: that everything is :heavy_check_mark: 
4. Merge the PR asap

sethrj commented:
Amazing! Thanks @alalazo !

adamjstewart commented:
Is the test the only thing that was preventing this PR from being merged? I was under the impression that there was a problem with the solution itself.

alalazo commented:
@adamjstewart The issue that @becker33 was raising is that we don't have a way to discriminate between an external and a binary package when `reuse:true`. This bug (which gives an inconsistent internal solution in ASP) don't allow _anything_ matching an external to be reused, so a user can say:
```yaml
mpi:
  buildable: false
packages:
  openmpi:
    externals:
    - spec: "openmpi@4.1.4"
```
and be sure it is using the external spec. I think with #27987 we can use hard-requirements to pin the required `virtual` into specs.

alalazo committed:
Test that users have a way to select a virtual

Currently, users have to resort to a workaround, i.e.
they have to put unsatisfiable requirements on reusable
packages they could otherwise pick.

This ought to be solved by extending the "require"
attribute to virtual packages, so that one can:
```yaml
mpi:
  require: 'multi-provider-mpi'
```

alalazo commented:
I added a new test. Currently users can select an external, by imposing unsatisfiable requirements on other providers that can be reused. This is not ideal, but works. If people agree (@adamjstewart @sethrj @becker33) I would merge the bug fix like it is for the time being, and then start working on extending the `packages:<pkg>:require` attribute to virtual packages. Ideally I'd like to have:
```yaml
packages:
  mpi:
    buildable: false
    require: 'openmpi@3.0.1'
  openmpi:
    externals:
    - spec: 'openmpi@3.0.1'
      prefix: /usr
```
applied whenever `mpi` is in the DAG as a virtual spec.

alalazo commented:
@becker33 Would https://github.com/spack/spack/pull/31558#issuecomment-1217867075 be fine with you?

psakievich commented:
@becker33 any updates on this?

alalazo commented:
For the record: I submitted https://github.com/spack/spack/pull/32369 today so I hope we'll converge to the syntax in https://github.com/spack/spack/pull/31558#issuecomment-1217867075 pretty soon (no matter whether this PR or #32369 gets merged first).

alalazo committed:
Use "require" instead of hack

alalazo commented:
#32369 has just been merged, and the second unit test updated accordingly

becker33 commented on lib/spack/spack/solver/concretize.lp:
```diff
@@ -464,7 +464,9 @@ version_weight(Package, Weight) :- external_version(Package, Version, Weight).
 
 % if a package is not buildable (external_only), only externals are allowed
-external(Package) :- external_only(Package), node(Package).
+external(Package) :- external_only(Package),
```
I still think we shouldn't be lying to Spack about whether it's an external. If it's a previously installed external, we should be marking that appropriately, not pretending it's not an external.

alalazo replied:
Imo the visible effect to users is consistent. The property is called `buildable` (not `external_only`), and the user can require constraints to be sure clingo selects a spec matching the external.

becker33 commented on lib/spack/spack/test/concretize.py:
```diff
@@ -1767,3 +1767,68 @@ def test_git_ref_version_errors_if_unknown_version(self, git_ref):
+        assert external3.dag_hash() == external1.dag_hash()
+
+    @pytest.mark.regression("31484")
+    def test_user_can_select_externals_with_require(self, mutable_database):
```
If I understand correctly, this only ensures we can force the provider, not the specific spec.

```
$ spack install mvapich2@2.3.7
$ spack config add packages:mvapich2:buildable:false
$ spack config edit  # add external for mvapich2@2.3.7
$ spack install mvapich2
```
would reuse the `mvapich2@2.3.7` installed on the first line, and there is no recourse to force it to use the external as far as I understand.

alalazo replied:
True. Then you can call your external `mvapich2@2.3.7.special`. I hope this will be a niche case

alalazo committed:
Prevent conflicts to be enforced on specs that can be reused.

alalazo committed:
Rename the "external_only" fact to "buildable_false", to better reflect its origin

alalazo committed:
Update documentation

alalazo commented:
@becker33 The bug occurring when:
1. A spec is installed
2. Later, a conflict is added for which the spec at 1 is a match
3. The spec installed at 1 cannot be reused

has been fixed in 30e8edc, and a unit test has been included to avoid regressions. 

I checked the rules in which `external(Package)` is involved and it seems to me they all come from before we introduced `reuse`. This is to say that external in `concretize.lp` means "defined as external in packages.yaml" since the `external(Package)` atom was introduced. One example for that is for instance:

https://github.com/spack/spack/blob/1c22af8ef4947d920b4d86e3571acdc2cdc8e377/lib/spack/spack/solver/concretize.lp#L450-L461

If a package is `external` we can only choose versions that are declared in `packages.yaml`. Not accounting for this, caused  the bug in #31484[^1]  and is fixed here. In 27adf2a I renamed an internal atom to adhere better to where it originates. In e569ba0 I updated docs to add some details on `buildable: false` and its behavior. Let me know how these changes looks.[^2]


[^1]: What gets installed is "frozen" and can get out of sync with a changing `packages.yaml`
[^2]: In the long term I hope we can get rid of most of this custom logic for externals, and that Spack can automatically detect packages and pull them into the local store in a reserved namespace (e.g. `spack.pkg.externals`), without the need for users to declare them manually. I don't think though implementing all that is necessary to fix this particular bug. Instead, I think this PR can solve a high-priority bug that has been open for a month and a half and give us time to consider how to design a better support for externals that blends with other requests to better tune what can be reused.

becker33 commented on lib/spack/docs/build_settings.rst:
```diff
@@ -121,9 +123,15 @@ be:
+   If ``concretizer:reuse`` is on (see :ref:`concretizer-options` for more information on that flag)
+   pre-built specs include specs already available from a local store, an upstream store, a registered
+   buildcache or specs marked as externals in ``packages.yaml``. If ``concretizer:reuse`` is off, only
+   external specs in ``packages.yaml`` are include d in the list of pre-built specs.
```
```suggestion
   external specs in ``packages.yaml`` are included in the list of pre-built specs.
```

alalazo committed:
Update comment

Co-authored-by: Greg Becker <becker33@llnl.gov>

adamjstewart commented:
Thanks for the hard work on this @alalazo and @becker33! It's unfortunate that it ended up being more involved than expected, but glad to see this bug finally fixed!

becker33 commented:
Give @alalazo all the credit, I just complained about solutions and he came up with new ones

adamjstewart commented:
Playing Devil's advocate and finding ways that a solution might break is just as important 😄
